### PR TITLE
BLD: Add pyproject.toml

### LINKED ISSUE #20775: Add pyproject.toml
Now that pip 19 is out, we can attempt to re-add pyproject.toml

---

The release of pip version 10 and the presence of a `pyproject.toml` in our latest released versions gives problems for installing pandas in the following cases:

**Problem with installing pandas for Python 3.4**

* You get the error *"Double requirement given: numpy ..."*
* The reason for this error is that pandas 0.21-0.22 no longer supports Python 3.4, and thus no longer distributes binary wheels for Python 3.4. Therefore, pip tries to install pandas from source, and this is what now started to fail with Pip version 10.
* **Solution**:
  * Install an older version of pandas

    ```
    pip install pandas<0.21
    ```

  * Workaround for package maintainers to specify the pandas version in `install_requires`: https://github.com/pandas-dev/pandas/issues/20697#issuecomment-381372004
* The fact that those version are downloaded for Python 3.4 (instead of an older compatible version) was a bug in our setup.py, and this will be fixed in future versions of pandas: https://github.com/pandas-dev/pandas/pull/20698

**Problem with installing pandas on "special" platforms (32bit, PyPy, ARM, ...) (building from source)**:

* You get the error *"No matching distribution found for Cython"*
* The reason for this error is that we do not provide wheels for such python versions or platforms. In such a case, `pip` will try to build pandas from source, but this started to fail with pip version 10. This is because their limited support for PEP518, and they require that all build dependencies of pandas are installed from wheels (https://github.com/pypa/pip/issues/5229), which are often also not available on those python versions / platforms.
* **Solution**
  * Use an older version of pip
  * Use `pip install pandas --no-build-isolation`
* See eg https://github.com/pandas-dev/pandas/issues/20666, https://github.com/pandas-dev/pandas/issues/20771

### PR DESCRIPTION
**Pull request by TomAugspurger**
* Adds pyproject.toml to support this.

* Bumps minimum Cython version to the version supporting Python 3.8.

Closes https://github.com/pandas-dev/pandas/issues/20775
xref https://github.com/pandas-dev/pandas/issues/27435 (check if closes)
xref https://github.com/pandas-dev/pandas/pull/25227, https://github.com/pandas-dev/pandas/pull/20718, https://github.com/pandas-dev/pandas/pull/16745

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+[build-system]
+# Minimum requirements for the build system to execute.
+requires = [
+    "setuptools",
+    "wheel",
+    "Cython>=0.29.13",  # Note: sync with setup.py
+    "numpy==1.13.3; python_version=='3.5' and platform_system!='AIX'",
```
Comment (TomAugspurger) [inline pyproject.toml]: These were copied from scipy. I'm not sure what the platform_system stuff is for.

File: MANIFEST.in [hunk 2]
```diff
@@ -3,14 +3,18 @@ include LICENSE
...
 
 graft doc
-prune doc/build
-
 graft LICENSES
-
 graft pandas
 
+prune doc/build
+prune asv_bench
```
Comment (TomAugspurger) [inline MANIFEST.in]: Any reason to include these? We could be even a bit stricter about what we include.

Maybe I should just revert these changes for now.

**Comment by jbrockmendel** [comment]
I take it there's no way to get the pyroject.toml stuff in setup.cfg?

**Comment by jbrockmendel** [comment]
should/could we test for the desired behavior?  or rare enough to check manually?

**Comment by TomAugspurger** [comment]
> I take it there's no way to get the pyroject.toml stuff in setup.cfg?

Nope.

> should/could we test for the desired behavior? or rare enough to check
manually?

We could use pip to compile a wheel from source and install that wheel into
our test environment? Dunno how much effort that will be.

**Comment by TomAugspurger** [comment]
I'm having trouble with editable installs. I'm need to pass the `--no-build-isolation` in order to use the extensions built with `python setup.py build_ext -i`.

File: MANIFEST.in (hunk 2 above)
Comment (WillAyd) [reply MANIFEST.in]: I think we need scripts for doctoring validation. Rest are probably OK to exclude

File: pyproject.toml (hunk 1 above)
Comment (WillAyd) [reply pyproject.toml]: Not sure if it matters but we technically required `3.5.3` as a minimum

File: MANIFEST.in (hunk 2 above)
Comment (TomAugspurger) [reply MANIFEST.in]: Note these are for the sdist. Do we care about people installing from the sdist, rather than GitHub, being able to run docstring validation?

Comment (WillAyd) [reply MANIFEST.in]: Maybe not. `pd.test()` would no longer have the the `scripts/tests` files available so probably OK so long as we don't set expectations elsewhere that those files should be run as part of a typical purest run

**Comment by TomAugspurger** [comment]
So from what I can tell, `pip install -e .` isn't going to play nicely with a previous `build_ext --inplace` (https://discuss.python.org/t/pip-19-1-and-installing-in-editable-mode-with-pyproject-toml/1553/68). So I guess the recommendation is to use `python setup.py develop` rather than `python -m pip install -e .`

Has anyone else followed this setup tools / pip issue? Does that sound right?

**Comment by jorisvandenbossche** [comment]
The editable install is a known issue (for me that is a reason to not yet use pyproject.toml on other projects, but for pandas the benefits are probably bigger than the problem with `pip install -e .`. 
Pip install will do build isolation once there is a pyproject.toml, which is something you exactly don't want (IMO) when doing an editable install. 
So as you say either passing a `--no-build-isolation`, or either `python setup.py develop`.

Can we do adding the pyproject.toml separately from the stop shipping cythonized files? Eg a separate PR as precursor for this (given all the previous problems I prefer to do this very explicitly and ) Or at least not hide it in a commit / PR that seems to be about something else.

**Comment by jorisvandenbossche** [comment]
> So as you say either passing a --no-build-isolation, or either python setup.py develop.

From a comment that I made earlier when running into this (https://issues.apache.org/jira/browse/ARROW-5210), it seems that I needed to do `pip install -e . --no-use-pep517 --no-build-isolation`. But I don't remember from the top of my head why both flags were needed.

**Comment by TomAugspurger** [comment]
> Can we do adding the pyproject.toml separately from the stop shipping cythonized files? Eg a separate PR as precursor for this (given all the previous problems I prefer to do this very explicitly and ) Or at least not hide it in a commit / PR that seems to be about something else.

Sure. I think Pyproject.toml should be first, and needs to be back ported. I'll do that here, since it's dominated the discussion.

**Comment by TomAugspurger** [comment]
Windows people, is it possible to do a glob that expands to a file in a `.bat` file? E.g. I have 

```python
bash-5.0$ ls wheelhouse
pandas-0.25.0+323.g55294ca9a.dirty-cp37-cp37m-macosx_10_14_x86_64.whl
bash-5.0$ pip install wheelhouse/*.whl
```

That puts the single `whl` file in the `pip install` command.

File: pyproject.toml (hunk 1 above)
Comment (jorisvandenbossche) [reply pyproject.toml]: Do we have any idea if pandas builds at all on AIX ?

> Not sure if it matters but we technically required 3.5.3 as a minimum

This line does not control the python version, it is only saying that if you have python 3.5, you need that numpy. 

But that said, I am wondering if there is a way to specify python_requires as well in the pyproject.toml world.

Comment (TomAugspurger) [reply pyproject.toml]: I suspect not, since this is invoked by pip, and pip is tied to a particular python?

Comment (TomAugspurger) [reply pyproject.toml]: @ayappanec do you happen to know if pandas can be built / used with AIX? We're trying to determine if these workarounds you added to SciPy's pyproject.toml are needed in ours as well.

Comment (jorisvandenbossche) [reply pyproject.toml]: But that pip with a specific python still needs to know if it can build/install pandas or not (depending on the python version). Currently this information is in setup.py, but I thought pyproject.toml would allow (if not using setuptools, but eg flit) having no setup.py. 

Anyway, that's not an important issue for this PR ;)

Comment (ayappanec) [reply pyproject.toml]: Numpy 1.14.0 ( and from that level onwards ) has a patch that is required for building projects like Scipy (which depends on numpy's script for gfortran flags) in AIX . So if Pandas also has the same approach , then we need to have these same conditions.

Comment (TomAugspurger) [reply pyproject.toml]: Thanks. Pandas has NumPy as a build dependency, but we don't do anything with fortran. So it sounds like we're safe to remove?

Comment (ayappanec) [reply pyproject.toml]: Okay. But still i feel like having these conditions doesn't do any harm i guess. So better we will have those on the safer side.

**Comment by TomAugspurger** [comment]
Just the SSL certificate doc failure.

File: pyproject.toml (hunk 1 above)
Comment (jorisvandenbossche) [reply pyproject.toml]: > Since we are targeting 1.0 now I guess can remove the line for 3.5 support

Master still has 3.5 support I think?

**Comment by jorisvandenbossche** [comment]
@TomAugspurger it might be that the `--no-use-pep517` is not needed, as they reverted some things in a pip bug fix release: https://pip.pypa.io/en/stable/news/#id50

File: pyproject.toml (hunk 1 above)
Comment (TomAugspurger) [reply pyproject.toml]: Yeah, I don't know that we've discussed dropping python 3.5 for 1.0. I think the idea was to follow the NEP policy if that gets done in time (I think it's supposed to be merged this week?)

**Comment by TomAugspurger** [comment]
Huh, I'll try again, but I recall needing it and I'm using pip 19.2.3 (that change was in 19.1.1)

File: pyproject.toml (hunk 1 above)
Comment (jreback) [reply pyproject.toml]: yeah, thought we should prob remove for 3.5

**Comment by jreback** [comment]
lgtm. merge when ready.

File: pyproject.toml (hunk 1 above)
Comment (simonjayhawkins) [reply pyproject.toml]: > Yeah, I don't know that we've discussed dropping python 3.5 for 1.0. I think the idea was to follow the NEP policy if that gets done in time (I think it's supposed to be merged this week?)

https://github.com/pandas-dev/pandas/issues/27492#issuecomment-513545165

**Comment by TomAugspurger** [comment]
> it might be that the --no-use-pep517 is not needed

Oh, I understand now. The `--no-use-pep517` is indeed not necessary. The `--no-build-isolation` *is* necessary to reuse previously built extensions.

**Comment by TomAugspurger** [comment]
All green.

As a reminder, I'll followup with a PR to clean up our sdist (including removing the cythonized files).

**Comment by TomAugspurger** [comment]
Oh, and just to make sure everyone is aware, @pandas-dev/pandas-core the proper way to rebuild extensions is now

```python
python setup.py build_ext -i  # -j 8 for parrallel
python -m pip install -e . --no-build-isolation
```

if you leave off the `--no-build-isolation` you'll be rebuilding from scratch each time.

**Comment by WillAyd** [comment]
@TomAugspurger just to be clear we have to run the `pip install -e` locally now when rebuilding instead of just `python setup.py build_ext --inplace`?

**Comment by TomAugspurger** [comment]
If you weren't previously using a local editable install then there's no change to your workflow.

**Comment by WillAyd** [comment]
Sounds good. We should probably update the contributing guide for that - I'll open an issue I think community will pick up

**Comment by WillAyd** [comment]
Nvm...missed already here - thanks for that

File: doc/source/development/contributing.rst [hunk 3]
```diff
@@ -208,7 +208,7 @@ We'll now kick off a three-step process:
 
    # Build and install pandas
    python setup.py build_ext --inplace -j 4
-   python -m pip install -e .
+   python -m pip install -e --no-build-isolation .
```
Comment (simonjayhawkins) [inline doc/source/development/contributing.rst]:
```
$ python -m pip install -e --no-build-isolation .
ERROR: --no-build-isolation is not a valid editable requirement. It should either be a path to a local project or a VCS URL (beginning with svn+, git+, hg+, or bzr+).
(pandas-dev)
simon@DESKTOP-F063QUA MINGW64 ~/OneDrive/code/pandas-simonjayhawkins (master)
$ python -m pip install -e . --no-build-isolation
Obtaining file:///C:/Users/simon/OneDrive/code/pandas-simonjayhawkins
    Preparing wheel metadata ... done
```

Comment (jorisvandenbossche) [reply doc/source/development/contributing.rst]: That's a typo, the `.` needs to come after the `-e` -> https://github.com/pandas-dev/pandas/pull/28445
